### PR TITLE
Set and get team active sprint

### PR DESCRIPTION
**Pull request by duanxianbo**
(task:15602)

### CODE REVIEW

**Comment by duanxianbo** [comment]
@magicoder10 it is ready for review

File: core/api/sprint_rpc.go [hunk 1]
```diff
@@ -31,7 +31,7 @@ func (s SprintRPC) Start(runner *runner.ServiceRunner) *errs.Error {
 }
 
 func (s SprintRPC) GetCurrentSprint(ct context.Context, req *proto.GetCurrentSprintRequest) (*proto.SprintMsg, error) {
-	sprint, err := s.sprintService.FindCurrentSprint(ct, req.TeamId)
+	sprint, err := s.sprintService.GetTeamActiveSprint(ct, req.TeamId)
```
Comment (naturetime10) [inline core/api/sprint_rpc.go]:
```suggestion
	sprint, err := s.sprintService.GetActiveSprint(ct, req.TeamId)
```

File: core/daov2/sprint.go [hunk 2]
```diff
@@ -16,6 +16,7 @@ type Sprint interface {
 	FindSprintsByIDsWithTx(ct context.Context, tx *transaction.Transaction, sprintIDs []uint64) ([]entity.Sprint, *errs.Error)
 	FindSprintsByTeamIDWithTx(ct context.Context, tx *transaction.Transaction, teamID uint64) ([]entity.Sprint, *errs.Error)
 	FindAllSprintsWithTx(ct context.Context, tx *transaction.Transaction) ([]entity.Sprint, *errs.Error)
+	FindActiveSprintByTeamIDWithTx(ct context.Context, tx *transaction.Transaction, teamID uint64) (entity.Sprint, *errs.Error)
```
Comment (naturetime10) [inline core/daov2/sprint.go]:
```suggestion
```

File: core/daov2/sqldb/team.go [hunk 3]
```diff
@@ -216,6 +216,23 @@ func (t Team) UpdateTeam(ct context.Context, tx *transaction.Transaction, team e
...
+		sprintID,
+		teamID,
+	)
+	if err != nil {
+		return errs.NewError(errs.Unknown, err.Error())
+	}
+
+	return nil
+}
+
```
Comment (naturetime10) [inline core/daov2/sqldb/team.go]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]
Let's make sure each Dao corresponds to 1 DB table and delegate business logic to service layer.

File: core/daov2/sqldb/sprint.go [hunk 4]
```diff
@@ -174,6 +174,39 @@ func (s Sprint) FindSprintsByTeamIDWithTx(ct context.Context, tx *transaction.Tr
...
+				"team active sprint not found: teamID=%v", teamID))
+	}
+
+	if err != nil {
+		return entity.Sprint{}, errs.NewError(errs.Unknown, err.Error())
+	}
+
+	return sprint, nil
+}
+
```
Comment (naturetime10) [inline core/daov2/sqldb/sprint.go]:
```suggestion
```

File: core/service/sprint.go [hunk 5]
```diff
@@ -84,47 +85,121 @@ func (s Sprint) FindSprints(ct context.Context, filter *SprintFilter) ([]entity.
...
+
+		if sprint.OwningTeamID != teamID {
+			return errs.NewError(errs.InvalidArgument, "Sprint does not belong to team")
+		}
+
+		team.ActiveSprintID = &sprintID
+		updatedAt := time.Now().UTC()
+		team.UpdatedAt = &updatedAt
+
+		updateTeamActiveSprintMutation := mutation.NewUpdateTeamActiveSprintMutation(
```
Comment (naturetime10) [inline core/service/sprint.go]: Should we use `update team mutation` instead?

File: core/service/sprint.go [hunk 6]
```diff
@@ -84,47 +85,121 @@ func (s Sprint) FindSprints(ct context.Context, filter *SprintFilter) ([]entity.
 	return sprints, nil
 }
 
-func (s Sprint) FindCurrentSprint(ct context.Context, teamID uint64) (entity.Sprint, *errs.Error) {
-	var sprint entity.Sprint
+func (s Sprint) GetActiveSprint(ct context.Context, teamID uint64) (*entity.Sprint, *errs.Error) {
```
Comment (naturetime10) [inline core/service/sprint.go:89]: Is this API actually needed for GraphQL API?

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/service/sprint.go (hunk 6 above)
Comment (duanxianbo) [reply core/service/sprint.go:89]: no, but it is needed in grpc service that is needed by github apps

File: core/service/sprint.go (hunk 5 above)
Comment (duanxianbo) [reply core/service/sprint.go]: I was thinking maybe we can make the payload less heavy, but actually the team has just a couple of more fields, will remove this and use update team mutation instead

File: core/realtime/collection.go [hunk 7]
```diff
@@ -17,4 +17,5 @@ const (
 	ClientCollectionType               CollectionType = "Client"
 	TaskActivityCollectionType         CollectionType = "TaskActivity"
 	SprintParticipantCollectionType    CollectionType = "SprintParticipant"
+	TeamActiveSprintCollectionType     CollectionType = "TeamActiveSprint"
```
Comment (naturetime10) [inline core/realtime/collection.go]:
```suggestion
```

File: core/service/sprint.go (hunk 5 above)
Comment (naturetime10) [reply core/service/sprint.go]: Let's use update Team mutation.

File: migrations/20230409101508_add_active_sprint_column.sql [hunk 8]
```diff
@@ -0,0 +1,7 @@
+-- +migrate Up
+ALTER TABLE team
+	ADD active_sprint_id BIGINT REFERENCES sprint (id) ON UPDATE CASCADE ON DELETE CASCADE;
```
Comment (naturetime10) [inline migrations/20230409101508_add_active_sprint_column.sql]:
```suggestion
	ADD active_sprint_id BIGINT REFERENCES sprint (id) 
	    ON UPDATE CASCADE 
	    ON DELETE CASCADE;
```

File: core/service/sprint_test.go [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+
+	testkit.StartServiceInstance(cloudTestKitConfig, virtualNetwork, cloudTestKit.ServiceInstanceRunner)
+
+	teamyPrometheus := metricstest.NewNoopMetrics()
+	cloudClientCfg := rpc.ConnectionConfig{
+		Host:          testkit.GRPCServerHost,
+		Port:          testkit.GRPCServerPort,
+		ShouldEncrypt: false,
+		GetAccessToken: func() string {
+			return "accessToken"
```
Comment (naturetime10) [inline core/service/sprint_test.go]: Need to generate this on the fly

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: core/service/sprint.go [hunk 10]
```diff
@@ -84,47 +86,123 @@ func (s Sprint) FindSprints(ct context.Context, filter *SprintFilter) ([]entity.
...
+	var team entity.Team
+	err := txCtx.withTransactions(false, func(tx *transaction.Transaction, rtTx *realtime.Transaction) *errs.Error {
+		var err *errs.Error
+		team, err = s.teamDaoV2.FindTeamByIDWithTx(ct, tx, teamID)
+		if err != nil {
+			return err
+		}
+
+		sprint, err := s.sprintDaoV2.FindSprintByIDWithTx(ct, tx, sprintID)
+
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
```

File: core/service/sprint.go [hunk 11]
```diff
@@ -84,47 +86,123 @@ func (s Sprint) FindSprints(ct context.Context, filter *SprintFilter) ([]entity.
...
+		}
+
+		if sprint.OwningTeamID != teamID {
+			return errs.NewError(errs.InvalidArgument, "Sprint does not belong to team")
+		}
+
+		team.ActiveSprintID = &sprintID
+		updatedAt := time.Now().UTC()
+		team.UpdatedAt = &updatedAt
+
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
```

File: core/service/sprint_test.go [hunk 12]
```diff
@@ -0,0 +1,528 @@
...
+	if !assert.Nil(t, internalErr) {
+		return SprintTestRef{}, false
+	}
+
+	apiToken, internalErr := servicetest.GetServiceAccountAPIToken(cloudTestKit.IdentityService)
+	if !assert.Nil(t, internalErr) {
+		return SprintTestRef{}, false
+	}
+
+	testkit.StartServiceInstance(cloudTestKitConfig, virtualNetwork, cloudTestKit.ServiceInstanceRunner)
```
Comment (naturetime10) [inline core/service/sprint_test.go]:
```suggestion
	testkit.StartServiceInstance(cloudTestKitConfig, virtualNetwork, cloudTestKit.ServiceInstanceRunner)
	
	apiToken, internalErr := servicetest.GetServiceAccountAPIToken(cloudTestKit.IdentityService)
	if !assert.Nil(t, internalErr) {
		return SprintTestRef{}, false
	}
```

File: core/service/sprint_test.go [hunk 13]
```diff
@@ -0,0 +1,528 @@
...
+			if testCase.prepareData != nil {
+				err := testCase.prepareData(sprintTestRef, testCase.requesterUserID)
+				if !assert.Nil(t, err) {
+					return
+				}
+			}
+
+			ct := context.Background()
+			ct = ctx.NewContextWithUserID(ct, testCase.requesterUserID)
+			now := time.Now()
```
Comment (naturetime10) [inline core/service/sprint_test.go]:
```suggestion
			now := time.Now().UTC()
```

File: core/service/sprint_test.go [hunk 14]
```diff
@@ -0,0 +1,528 @@
...
+	}
+
+	apiToken, internalErr := servicetest.GetServiceAccountAPIToken(cloudTestKit.IdentityService)
+	if !assert.Nil(t, internalErr) {
+		return SprintTestRef{}, false
+	}
+
+	testkit.StartServiceInstance(cloudTestKitConfig, virtualNetwork, cloudTestKit.ServiceInstanceRunner)
+
+	teamyPrometheus := metricstest.NewNoopMetrics()
```
Comment (naturetime10) [inline core/service/sprint_test.go]:
```suggestion
	noopMetrics := metricstest.NewNoopMetrics()
```

File: core/service/sprint_test.go [hunk 15]
```diff
@@ -0,0 +1,528 @@
...
+				CreatedAt:     now,
+			}
+
+			err = sprintTestRef.teamDaoV2.CreateTeam(ct, tx, team)
+			if !assert.Nil(t, err) {
+				return
+			}
+
+			updatedTeam, internalErr := sprintTestRef.sprintService.SetTeamActiveSprint(ct, teamID, sprintID1)
+			if testCase.expectedErr != nil {
```
Comment (naturetime10) [inline core/service/sprint_test.go:518]:
```suggestion
			if testCase.expectedSprint1Err != nil {
```

File: core/service/sprint_test.go [hunk 16]
```diff
@@ -0,0 +1,528 @@
...
+				assert.Equal(t, testCase.expectedErr.Code, internalErr.Code)
+				return
+			} else if !assert.Nil(t, internalErr) {
+				return
+			}
+
+			assert.Equal(t, sprintID1, *updatedTeam.ActiveSprintID)
+
+			updatedTeam, internalErr = sprintTestRef.sprintService.SetTeamActiveSprint(ct, teamID, sprintID2)
+			if testCase.expectedErr != nil {
```
Comment (naturetime10) [inline core/service/sprint_test.go:528]:
```suggestion
			if testCase.expectedSprint2Err != nil {
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Comment by duanxianbo** [comment]
@magicoder10

File: core/service/sprint.go [hunk 17]
```diff
@@ -84,47 +86,123 @@ func (s Sprint) FindSprints(ct context.Context, filter *SprintFilter) ([]entity.
...
+		updateTeamMutation := mutation.NewUpdateTeam(
+			s.logger,
+			s.stateSyncer,
+			s.teamDao,
+			s.teamDaoV2,
+			team,
+		)
+
```
Comment (naturetime10) [inline core/service/sprint.go]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**